_docs/Qabel-Protocol-Box.md

```diff
@@ -0,0 +1,578 @@
+
+A set of protocols to store files and folders on a VOLUME that is on AWS S3 and managed by a third party who doesn't need to be trusted.
+
+## Used services
+
+Qabel Box uses an Accounting server which controls the access to AWS S3. Every client who needs write access has to be authenticated by the server and then receives a set of credentials for direct access to the VOLUME.
+Qabel Box also directly uses AWS S3 to store the blocks and metadata.
+
+## Accounting server
+
+The accounting server controls write access to the S3 bucket. Registered users (not identities!) can request temporary credentials for AWS.
+All data is sent as JSON and UTF-8.
+
+### Required permissions for the AWS user
+The accounting server uses an AWS user to create the temporary credentials. He must have the following permissions:
+
+* "sts:GetFederationToken" to generate the credentials
+* "s3:GetObject" for read access
+* "s3:PutObject" for uploading
+* "s3:ListBucket" is used to find unreferenced files (not implemented)
+
+These permissions should only be granted for the Qabel box S3 bucket.
+
+### Login
+The login method grants a new authentication token.
+
+* Ressource: /api/v0/auth/login
+* Method: POST
+* Request data: `{username: STR, password: STR}`
+* Response data: `{key: STR}`
+
+The authentication token is used by including the header "Authorization" with the value "Token " concatenated with the key.
+
+For example: `Authorization: Token 70373def6f3766ab1782700cba4404`
+
+For every method except login, this authorization header is required.
+
+### Token
+The token ressource controls the federation token.
+
+* Ressource: /api/v0/token
+* Method: POST
+* Request data: None
+* Response data: `{AccessKeyId: STR, SecretAccessKey: STR, SessionToken: STR}`
+
+### Prefix
+The prefix ressource controls all prefixes of the user. All prefixes are fully qualified URLs.
+
+Create a new prefix:
+
+* Ressource: /api/v0/prefix
+* Method: POST
+* Request data: None
+* Response data: `{prefix: STR}`
+
+Get a list of available prefixes
+
+* Ressource: /api/v0/prefix
+* Method: GET
+* Request data: None
+* Response data: `{prefixes: [STR]}`
+
+## Structure of a VOLUME
+
+A Volume consists of metadata files and blocks. Every VOLUME has a metadata file at VOLUME/\<index\> which is the starting point and contains references to other objects. All file names on S3 are UUIDs.
+
+All mtime values are seconds since epoch in UTC. Blocks are stored at VOLUME/blocks/.
+
+The metadata file stores information equivalent of this example JSON document, but stored in an SQLite database (the database schema is explained later):
+
+```
+{
+root: "https://qabelbox.s3.amazonaws.com/users/b5911736-9ace-a799-8e34-dd9c17acff9a/",
+name: "index",
+spec_version: 0,
+version: {version: "85bc5ead74c52df59c3abd3340ff9d6bd821acd61189950aec4f68c37b773a20",
+          time:  1445963627},
+last_change_by: "487a481f-4d93-cef0-4475-88ee576d37fd",
+shared: [
+	"aa8c3f39-edc5-00b0-ab8b-ba66d05b60db" : { read: [
```

Comment:
JSON does not support comments. To make the examples valid JSON documents, I'd have to strip them.